source/common/secret/sds_api.cc

```diff
@@ -45,28 +51,35 @@ void SdsApi::resolveDataSource(const FileContentMap& files,
 }
 
 void SdsApi::onWatchUpdate() {
-  // Obtain a stable set of files. If a rotation happens while we're reading,
-  // then we need to try again.
-  uint64_t prev_hash = 0;
-  FileContentMap files = loadFiles();
-  uint64_t next_hash = getHashForFiles(files);
-  const uint64_t MaxBoundedRetries = 5;
-  for (uint64_t bounded_retries = MaxBoundedRetries; next_hash != prev_hash && bounded_retries > 0;
-       --bounded_retries) {
-    files = loadFiles();
-    prev_hash = next_hash;
-    next_hash = getHashForFiles(files);
-  }
-  if (next_hash != prev_hash) {
-    ENVOY_LOG_MISC(warn,
-                   "Unable to atomically refresh secrets due to > {} non-atomic rotations observed",
-                   MaxBoundedRetries);
-  }
-  const uint64_t new_hash = next_hash;
-  if (new_hash != files_hash_) {
-    resolveSecret(files);
-    update_callback_manager_.runCallbacks();
-    files_hash_ = new_hash;
+  // Filesystem reads and update callbacks can fail if the key material is missing or bad. We're not
+  // under an onConfigUpdate() context, so we need to catch these cases explicitly here.
+  try {
```

Comment:
Is it possible to have the scope of the try/catch be a bit smaller? Especially not the callbacks? NBD if not but just a thought.

Reply:
It needs to cover the callbacks, since they may throw exceptions, and the rotation happens outside of an `onConfigUpdate` when it's safe to do so.

Reply:
OK makes sense.